src/display/editor/stamp.js

```diff
@@ -0,0 +1,411 @@
+/* Copyright 2022 Mozilla Foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { AnnotationEditor } from "./editor.js";
+import { AnnotationEditorType } from "../../shared/util.js";
+import { StampAnnotationElement } from "../annotation_layer.js";
+
+/**
+ * Basic text editor in order to create a FreeTex annotation.
+ */
+class StampEditor extends AnnotationEditor {
+  #bitmap = null;
+
+  #bitmapId = null;
+
+  #bitmapPromise = null;
+
+  #bitmapUrl = null;
+
+  #canvas = null;
+
+  #observer = null;
+
+  #resizeTimeoutId = null;
+
+  static _type = "stamp";
+
+  constructor(params) {
+    super({ ...params, name: "stampEditor" });
+    this.#bitmapUrl = params.bitmapUrl;
+  }
+
+  static initialize(_l10n) {}
+
+  static updateDefaultParams(_type, _value) {}
+
+  static get defaultPropertiesToUpdate() {
+    return [];
+  }
+
+  get propertiesToUpdate() {
+    return [];
+  }
+
+  #getBitmap() {
+    if (this.#bitmapId) {
+      this._uiManager.imageManager.getFromId(this.#bitmapId).then(data => {
+        if (!data) {
+          this.remove();
+          return;
+        }
+        this.#bitmap = data.bitmap;
+        this.#createCanvas();
+      });
+      return;
+    }
+
+    if (this.#bitmapUrl) {
+      const url = this.#bitmapUrl;
+      this.#bitmapUrl = null;
+      this.#bitmapPromise = this._uiManager.imageManager
+        .getFromUrl(url)
+        .then(data => {
+          this.#bitmapPromise = null;
+          if (!data) {
+            this.remove();
+            return;
+          }
+          ({ bitmap: this.#bitmap, id: this.#bitmapId } = data);
+          this.#createCanvas();
+        });
+      return;
+    }
+
+    const input = document.createElement("input");
+    input.type = "file";
+    input.accept = "image/*";
+    this.#bitmapPromise = new Promise(resolve => {
+      input.addEventListener("change", async () => {
+        this.#bitmapPromise = null;
+        if (input.files.length === 0) {
```

Comment:
Looking at similar code in the GENERIC viewer, should we perhaps guard this as follows to prevent any issues?
```suggestion
        if (!input.files || input.files.length === 0) {
```